File: frontend/src/types/agent.ts
// types/agent.ts
// Agent 任务台相关类型，对齐后端 schemas/agent.py 与 cc/AGENT_SSE_EVENTS.md。

// --------------------------------------------------------------------------- //
// 出参（对齐 AgentSessionOut / AgentMessageOut / AgentApprovalOut）
// --------------------------------------------------------------------------- //
export interface AgentSession {
  id: number
  task_type: string | null
  title: string | null
  status: string
  runner_type: string
  user_id: number | null
  claude_session_id: string | null
  cost_usd: number | null
  error: string | null
  created_at: string | null
  updated_at: string | null
}

export interface AgentMessage {
  id: number
  session_id: number | null
  seq: number | null
  role: string | null // 'assistant' | 'tool' | 'user' | ...
  content: string | null
  tool_name: string | null
  tool_use_id: string | null
  tool_input: any | null
  tool_result: any | null
  created_at: string | null
}

export interface AgentApproval {
  id: number
  session_id: number | null
  sql: string
  danger_reason: string | null
  affected_estimate: number | null
  status: string // pending / approved / rejected / auto_executed / timeout / ...
  decided_by: number | null
  decided_at: string | null
  created_at: string | null
}

export interface SessionDetail {
  session: AgentSession
  messages: AgentMessage[]
  pending_approvals: AgentApproval[]
}

export interface TaskType {
  task_type: string
  title: string
}

// --------------------------------------------------------------------------- //
// SSE 事件（对齐 cc/AGENT_SSE_EVENTS.md，11 种 kind）
// --------------------------------------------------------------------------- //
export interface AgentEventHistory {
  kind: 'history'
  role: 'assistant' | 'tool'
  seq: number
  content: string | null
  tool_name: string | null
  tool_use_id: string | null
  tool_input: object | null
  tool_result: any | null
}

export interface AgentEventHistoryEnd {
  kind: 'history_end'
  status: string
}

export interface AgentEventTextDelta {
  kind: 'text_delta'
  text: string
}

export interface AgentEventToolUse {
  kind: 'tool_use'
  tool_name: string
  tool_input: object
  tool_use_id: string
  seq: number
}

export interface AgentEventToolResult {
  kind: 'tool_result'
  tool_use_id: string
  tool_result: any
  seq: number
}

export interface AgentEventDone {
  kind: 'done'
  is_error: boolean | null
  cost_usd: number | null
  permission_denials: string[] | null
  error: string | null
  synthesized?: boolean
  status?: string
}

export interface AgentEventError {
  kind: 'error'
  error: string
  is_error: boolean
}

export interface AgentEventApprovalNeeded {
  kind: 'approval_needed'
  approval_id: number
  sql: string
  danger_reason: string
}

export interface AgentEventSqlExecuted {
  kind: 'sql_executed'
  approval_id: number
  sql: string
  affected: number | null
  status: string // auto_executed / approved
  auto: boolean
  error: string | null
}

export interface AgentEventSqlRejected {
  kind: 'sql_rejected'
  approval_id: number
  sql: string
  status: 'rejected'
}

export interface AgentEventSqlTimeout {
  kind: 'sql_timeout'
  approval_id: number
  sql: string
}

export type AgentEvent =
  | AgentEventHistory
  | AgentEventHistoryEnd
  | AgentEventTextDelta
  | AgentEventToolUse
  | AgentEventToolResult
  | AgentEventDone
  | AgentEventError
  | AgentEventApprovalNeeded
  | AgentEventSqlExecuted
  | AgentEventSqlRejected
  | AgentEventSqlTimeout

File: backend/app/schemas/agent.py
"""Agent 任务台 API Pydantic schemas（Task 5）。

字段尽量对齐 ``AgentSession`` / ``AgentMessage`` / ``AgentApproval`` 模型，
datetime 统一序列化为 ISO 字符串。
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# --------------------------------------------------------------------------- #
# 入参
# --------------------------------------------------------------------------- #
class CreateSessionIn(BaseModel):
    """建会话入参。

    task_type 必须命中 ``TASK_TEMPLATES``（见 ``services/agent/task_templates.py``）；
    未知 task_type 由 ``agent_api.create_session`` 返回 400。可用类型清单见
    ``GET /api/v1/agent/task-types``。
    """

    task_type: str = Field(..., description="任务类型（fill_piece_weight 等）")
    force: bool = Field(False, description="强制重新处理全部（追加到 prompt 提示 Agent 忽略已处理标记）")


class PostMessageIn(BaseModel):
    """插话入参：在已终态会话上起 resume 新轮。"""

    text: str = Field(..., description="追加的用户消息文本")


class ApprovalDecisionIn(BaseModel):
    """审批决策入参。"""

    approved: bool = Field(..., description="True=批准并执行，False=拒绝")


# --------------------------------------------------------------------------- #
# 出参
# --------------------------------------------------------------------------- #
class AgentSessionOut(BaseModel):
    id: int
    task_type: Optional[str] = None
    title: Optional[str] = None
    status: str
    runner_type: str
    user_id: Optional[int] = None
    claude_session_id: Optional[str] = None
    cost_usd: Optional[float] = None
    error: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AgentMessageOut(BaseModel):
    id: int
    session_id: Optional[int] = None
    seq: Optional[int] = None
    role: Optional[str] = None
    content: Optional[str] = None
    tool_name: Optional[str] = None
    tool_use_id: Optional[str] = None
    tool_input: Optional[Any] = None
    tool_result: Optional[Any] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AgentApprovalOut(BaseModel):
    id: int
    session_id: Optional[int] = None
    sql: str
    danger_reason: Optional[str] = None
    affected_estimate: Optional[int] = None
    status: str
    decided_by: Optional[int] = None
    decided_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class SessionDetailOut(BaseModel):
    session: AgentSessionOut
    messages: list[AgentMessageOut]
    pending_approvals: list[AgentApprovalOut]
